### ISSUE TITLE
`GRPOTrainer` having different reward functions for training vs. evaluation

### ISSUE THREAD

**Issue by jamesbraza** [huggingface/trl#3018]
### Feature request

The request is to be able to provide a different reward function for training vs evaluation to `GRPOTrainer`.

### Motivation

It can be useful to specify a different reward function for training and evaluation. For example, one can train with dense rewards (aka `float` in [0, 1]) and evaluate with binary (just `0` or `1`) rewards.

### Your contribution

Happy to add an `eval_reward_funcs` with testing to `GRPOTrainer`.

**Comment by tchang1997** [comment]
+1 — As a hack, I've been getting around this by defining new reward functions and setting `reward_weight` to zero (so it still gets logged, but doesn't affect the "actual" reward).

**Comment by qgallouedec** [comment]
Hey @jamesbraza, sorry for the very slow reply, we're catching up on the backlog.

The use case makes sense, but `eval_reward_funcs` isn't a flag, it's a second reward pipeline: its own processing classes and weights, the DeepSpeed/FSDP prep duplicated for eval-only reward models, and mode-branching through `_calculate_rewards`. That's too much permanent surface for something that only ever touches logged metrics.

Good news: a weight of `0.0` already gets you there. Zero-weighted reward functions are still computed and logged per-mode, but contribute nothing to the advantages:

```python
def dense_reward(completions, **kwargs):
    return [similarity(c) for c in completions]  # drives training

def binary_reward(completions, **kwargs):
    return [float(is_correct(c)) for c in completions]  # eval metric only

trainer = GRPOTrainer(
    model="Qwen/Qwen3-0.6B",
    reward_funcs=[dense_reward, binary_reward],
    args=GRPOConfig(reward_weights=[1.0, 0.0], eval_strategy="steps", eval_steps=50),
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
)
```

You get `eval_rewards/binary_reward/mean` at eval, and training is driven purely by `dense_reward`.

One caveat: `binary_reward` also runs during training, free for a cheap verifier, not free if it's a reward *model* or something big to run.

Closing, but happy to reopen if you hit a wall. Thanks @tchang1997 for surfacing the same trick!
